Add mqtt for idf

martgras opened the pull request:
# What does this implement/fix? 

Adds mqtt support for esp-idf. 
- add support for TLS in idf
- refactor mqtt client



## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1601

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1776

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
```
esphome:
  name: testing

esp32:
  board: esp32dev
  framework:
    type: esp-idf

logger:
  level: VERBOSE

wifi:
  ssid: !secret wifi_sid
  password: !secret wifi_password

mqtt:
  broker: 192.168.1.114
  port: 1883
  discovery: true
  discovery_prefix: homeassistant
  on_message:
    topic: testing/sensor/uptime_sensor/state
    qos: 0
    then:
      - lambda: |-
          ESP_LOGD("Mqtt Test","testing/sensor/uptime_sensor/state=[%s]",x.c_str());

sensor:
  - platform: uptime
    name: Uptime Sensor
    id: uptime_sensor
    update_interval: 10s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

## Todo:
  - [x] SSL will be added in a follow-up PR. - implemented
  - [x] Clean-up.

misery commented:
Nice, I like to see mqtt with esp-idf. I tried your branch here and get an error.

```
Compiling .pioenvs/testing/src/esphome/components/mqtt/mqtt_component.o
In file included from src/esphome/components/mqtt/mqtt_client.cpp:1:
src/esphome/components/mqtt/mqtt_client.h:12:10: fatal error: AsyncMqttClient.h: No such file or directory

*************************************************************************
* Looking for AsyncMqttClient.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:AsyncMqttClient.h"
* Web  > https://platformio.org/lib/search?query=header:AsyncMqttClient.h
*
*************************************************************************

 #include <AsyncMqttClient.h>
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pioenvs/testing/src/esphome/components/mqtt/mqtt_client.o] Error 1
```

martgras commented:
That doesn't seem to be the correct code or maybe git didn't delete the files I remaned .  mqtt_client.h and mqtt_client.cpp have been renamed to mqttclient. You shouldn't have mqtt_client.* in the directory anymore

misery commented:
> That doesn't seem to be the correct code or maybe git didn't delete the files. I renamed mqtt_client.h and mqtt_client.cpp to mqttclient. You shouldn't have mqtt_client.* in the directory anymore

Yeah, I thought that, too. I multiple deleted the whole build directory and used a fresh virtualenv for esphome.

martgras committed:
add mqtt for idf

add common mqtt base class
add support for mqtt with idf
TLS support for idf using CA Certificate

Co-Authored-By: Flaviu Tamas <me@flaviutamas.com>

martgras commented:
@misery I updated the PR based on the latest dev changes 


For testing you can use:
```yaml
external_components:
  - source: github://pr#2979
    components: ["json"]
  - source: github://pr#2930
    components: ["mqtt"]
```
try esphome clean  <yaml> if see compilation errors about missing includes
Also make sure you are using the dev branch and it is up to date

misery commented:
Thanks! I tried external_component and had an include problem again. So I cleaned anything again and cherry-picked the changes to the dev branch. That worked! Maybe dirty working directory or something like this even it was a fresh installation. But I have ESP32-S2 with idf mqtt now! Yeah!

It looks a little bit slow in logging. It take one seconds for each line. But it finally works. :-)

martgras commented:
Interesting I even see a larger delay (~20 seconds sometimes). 
I just added  the  `idf_send_async` option to mqtt .  If true (default)  publish happens on the mqtt task and will be delayed 
If false publish will happen immediately  but thread can be blocked for up to 10 seconds (network timeouts)

If you want to try add `refresh: 1h` to external_components otherwise it will take a day before your client syncs again with github

```
external_components:
  - source: github://pr#2979
    components: ["json"]
  - source: github://pr#2930
    refresh: 1h
    components: ["mqtt"]
mqtt:
  broker: 192.168.1.114
  port: 1883
  discovery: true
  discovery_prefix: homeassistant
  idf_send_async: false
  on_message:
    topic: testing/sensor/uptime_sensor/state
    qos: 0
    then:
      - lambda: |-
          ESP_LOGD("Mqtt Test","testing/sensor/uptime_sensor/state=[%s]",x.c_str());
```

flaviut commented on esphome/components/mqtt/mqtt_client_idf.cpp:
```diff
@@ -0,0 +1,150 @@
+  mqtt_cfg_.buffer_size = MQTT_BUFFER_SIZE;
+  ESP_LOGV(TAG, "Trying to connect to: %s port:%d", mqtt_cfg_.host, this->mqtt_cfg_.port);
+
+  if (this->username_.empty()) {
```
things seem to work fine, which confuses me, since this only sets the username if there's nothing in it.

If this isn't supposed to do anything, it ought to be deleted, otherwise can you drop in a comment explaining how it works?

martgras replied:
fixed

flaviut commented on esphome/components/mqtt/mqtt_client_idf.cpp:
```diff
@@ -0,0 +1,150 @@
+      break;
+    case MQTT_EVENT_DISCONNECTED:
+      ESP_LOGV(TAG, "MQTT_EVENT_DISCONNECTED");
+      // TODO is there a way to get the disconnect reason?
```
There is not :(

https://github.com/espressif/esp-mqtt/blob/026253d802473ac9afd4cc747e2632e96f71acbb/mqtt_client.c#L700-L712

martgras replied:
ok thanks

flaviut commented on esphome/components/mqtt/mqtt_client_idf.cpp:
```diff
@@ -0,0 +1,150 @@
+      properties.retain = false;
+      properties.dup = false;
+      auto data_len = event->data_len;
+      if (data_len == 0)
```
Did you run into problems with testing here? If not, we should trust `data_len` if it is zero, since it is always set: https://github.com/espressif/esp-mqtt/blob/026253d802473ac9afd4cc747e2632e96f71acbb/mqtt_client.c#L996

In any case, we shouldn't use strlen, we should be defensive & bail out, rather than potentially working with corrupt data.

martgras replied:
removed it

flaviut reviewed (changes requested):
Generally looks good to me! I'm very happy to see CA support here, and thanks for doing the work.

flaviut commented:
> @oxan said in https://discord.com/channels/429907082951524364/524177279270780928/922595211484491816
>
> i think it might be cleaner if we just define our own interface for an MQTT client, and create implementations based upon the Arduino lib and IDF though, instead of mimicking the Arduino lib with an IDF implementation

I'm mildly against using a different interface than the Arduino one. The existing interface isn't perfect--the issues I see are:

- the use of raw strings in https://github.com/esphome/esphome/pull/2930/files#diff-a6f0409da74bdf8195008586e72677896ee42155f02e40efb02197648a23e556R42
- extra QOS parameter in https://github.com/esphome/esphome/pull/2930/files#diff-a6f0409da74bdf8195008586e72677896ee42155f02e40efb02197648a23e556R39
- the use of a different method naming convention than ESPHome.

But I don't feel like those changes outweigh the minuscule advantages of keeping the interface the same (mostly familiarity).

It's your project so your decision, but I'd be OK with the way things are here if this was my project.

oxan commented:
My biggest problem with imitating the Arduino interface is that it's bound to get ugly when the featureset isn't equal. If Arduino implements a feature we don't, we need to provide an empty method that does nothing, which is a great way for a future developer to lose an hour wondering why his code that looks perfectly fine doesn't work. If IDF has a feature that Arduino doesn't, we can't add that to the Arduino class. Using the Arduino interface also creates the expectation that the Arduino library is used, which again might be a surprise for future developers looking at this.

The other thing is that the Arduino interfaces usually aren't great, to say it diplomatically. That also caused that comment, but now that I've actually looked at the interface, it's not _that_ bad. That helpers implementing our own interface as well though, since it can mostly be a 1:1 translation.

That said, I don't think creating a wrapper interface will be that much work and that it'll be worth it, but I don't feel _too_ strongly about it. Having it like this is probably still better than having nothing on IDF.

oxan commented:
Looking at the PR itself, do we actually need to rename `mqtt_client.h`? I would expect that `#include <mqtt_client.h>` would always pick the IDF one as our own directory isn't on the bracketed-include search path. It's just a bit ugly to have that one file follow a different naming convention.

flaviut commented:
> I would expect that #include <mqtt_client.h> would always pick the IDF one as our own directory isn't on the bracketed-include search path.

I'd find that confusing. C++ has enough weird things and my brain is already full enough that I don't want to remember the difference between `""` and `<>`.

> That helpers implementing our own interface as well though, since it can mostly be a 1:1 translation.
>
> That said, I don't think creating a wrapper interface will be that much work and that it'll be worth it, but I don't feel too strongly about it.

I did some work on @martgras's wrapper PR at https://github.com/flaviut/esphome/tree/mqtt-refactor. I don't think the code there is any better, it still has to deal with differences between the ESP-IDF and Arduino versions. For example, the "rep" parameter, which does nothing for ESP-IDF.

Of the 3 issues I mentioned with the current API, the wrapper is able to fix two. The first one can be fixed with `std::string_view`, but it looks like we're using C++11, so that can't be done.

We also still need to know the concrete type of the MQTT provider. The arduino interface and the ESP-IDF interface use totally different methods of doing MQTTS. Doing a wrapper doesn't save us from that.

But anyway, take a look at the other branch and see what you think of it!

OttoWinter commented:
> I would expect that `#include <mqtt_client.h>` would always pick the IDF one as our own directory isn't on the bracketed-include search path.

Unfortunately the `esphome` dirs _are_ included in the bracketed-include search paths as well (see for example generated compile_commands.json). That's an artifact of the weird way we use the build system in repo setup.

OttoWinter commented on esphome/components/mqtt/__init__.py:
```diff
@@ -39,10 +40,18 @@
 DEPENDENCIES = ["network"]
-AUTO_LOAD = ["json", "async_tcp"]
+
+if KEY_CORE in CORE.data and CORE.using_esp_idf:
```
This is not going to work unfortunately.

It does work when just running `esphome run ...`, but with this the dashboard validation would be broken.

The dashboard validation keeps the modules around after they've been imported (so as to not having to reload them each time you type in a character in the editor, which would make it very slow). But if you switch to another file the modules are still loaded, but with a different config.

So we can't access any `CORE.` values during module loading, only in validation functions because those are re-run every time.

martgras replied:
Can I just remove the check and keep async_tcp in AUTO_LOAD
The only downside is that it's copied to the build  path.  Correct? 
Edit  - nope fails during validation - don't know how to address this

OttoWinter replied:
Not 100% sure, but I think the `async_tcp` component itself validates that it can only be loaded in arduino environments (as the libraries it uses are only compatible with arduino)

martgras replied:
Moved adding async_tcp to `to_code()`. 
I am not happy hardcoding library versions here  but since `AsyncMqttClient-esphome` is already hardcoded it doesn't make things worse IMHO :)

oxan replied:
It seems like we can remove the explicit AsyncTCP library dependencies as `AsyncMqttClient-esphome` depends on them, so PlatformIO adds them automatically.

OttoWinter commented on esphome/components/mqtt/__init__.py:
```diff
@@ -300,7 +315,17 @@ async def to_code(config):
+    # esp-idf only
+    if CONF_CERTIFICATE_AUTHORITY in config:
+        cg.add(var.set_ca_certificate(config[CONF_CERTIFICATE_AUTHORITY]))
+        cg.add(var.set_skip_cert_cn_check(config[CONF_SKIP_CERT_CN_CHECK]))
```
This key is optional in the schema. If you only set `CONF_CERTIFICATE_AUTHORITY` in the config but don't set `CONF_SKIP_CERT_CN_CHECK`, it would pass validation but fail with an exception here. Either make it so both have to be set together, or set defaults for this value.

OttoWinter commented on esphome/components/mqtt/__init__.py:
```diff
@@ -300,7 +315,17 @@ async def to_code(config):
+        add_idf_sdkconfig_option("CONFIG_MBEDTLS_HARDWARE_MPI", False)
+
+    if CONF_IDF_SEND_ASYNC in config and config[CONF_IDF_SEND_ASYNC]:
+        cg.add_define("IDF_MQTT_USE_ENQUEUE")
```
We usually name defines `USE_<COMPONENT_NAME>_<SUFFIX>` to more easily identify them as defines and see where they're from.

OttoWinter commented on esphome/components/mqtt/mqtt_client_base.h:
```diff
@@ -0,0 +1,111 @@
+  bool retain;
+};
+
+class MqttClientBase {
```
Inconsistent naming, we should either go for `MQTT` or ´Mqtt`, personally I'd be for `MQTT` as that's what we've used in the past too (I think)

OttoWinter commented on esphome/components/mqtt/mqtt_client_idf.cpp:
```diff
@@ -0,0 +1,149 @@
+    case MQTT_EVENT_PUBLISHED:
+      ESP_LOGV(TAG, "MQTT_EVENT_PUBLISHED, msg_id=%d", event->msg_id);
+      for (auto &callback : this->on_publish_) {
+        callback((int) event->msg_id);
```
From what thread is the IDF event handler called? If it's not from the main thread (which I think is the case), then we'd need to post these events into a queue and run the callbacks in the main thread. Otherwise we end up with lots of issues as ESPHome's code expects to be run from a single thread, and there's no synchronization.

martgras replied:
Depends - see https://github.com/espressif/esp-mqtt/blob/b86d42c130ac64a916ce6cf299d99f9756692394/mqtt_client.c#L1565

But I implemented a queue and use a semaphore  for synchronization

martgras committed:
implement CR suggestions

flaviut commented on esphome/components/mqtt/mqtt_client_idf.h:
```diff
@@ -0,0 +1,167 @@
+  static const size_t MQTT_BUFFER_SIZE = 4096;
+
+  void set_keep_alive(uint16_t keep_alive) final { this->keep_alive_ = keep_alive; }
+  void set_client_id(const char *client_id) override { this->client_id_ = client_id; }
```
No big deal, but I used `final` elsewhere to assist the compiler in devirtualizing these method calls.

martgras replied:
Thanks

martgras committed:
mark virtual methods final remove duplicate code

martgras committed:
format

martgras committed:
fix format after clang tidy

martgras committed:
add nolint to break clang-tidy loop

martgras committed:
fix nolint

martgras committed:
add nolint to header as well

oxan commented:
> I'd find that confusing. C++ has enough weird things and my brain is already full enough that I don't want to remember the difference between `""` and `<>`.

`<>` is for system libraries and `""` is for our own code. I find that a clearer and more explicit distinction than the filename having an underscore or not.

> Unfortunately the `esphome` dirs _are_ included in the bracketed-include search paths as well (see for example generated compile_commands.json). That's an artifact of the weird way we use the build system in repo setup.

The root directory might be, but I don't think `esphome/components/mqtt` is?

martgras committed:
handle event from loop

martgras committed:
remove hardcoded lib
clang tidy forever

martgras committed:
rename mqtt classes
add test

oxan committed:
Minor code review changes

oxan committed:
Convert MQTTBackend to use boolean return

oxan committed:
Rename mqtt_client_ member to mqtt_backend_

oxan committed:
Use CallbackManager

oxan commented on esphome/components/mqtt/mqtt_backend.h:
```diff
@@ -0,0 +1,82 @@
+namespace esphome {
+namespace mqtt {
+
+enum class MQTTEvent {
+  MQTT_EVENT_ANY = -1,
+  MQTT_EVENT_ERROR = 0,
+  MQTT_EVENT_CONNECTED,
+  MQTT_EVENT_DISCONNECTED,
+  MQTT_EVENT_SUBSCRIBED,
+  MQTT_EVENT_UNSUBSCRIBED,
+  MQTT_EVENT_PUBLISHED,
+  MQTT_EVENT_DATA,
+  MQTT_EVENT_BEFORE_CONNECT,
+  MQTT_EVENT_DELETED,
+};
```
This enum seems to be unused, can we delete it?

martgras replied:
done

oxan commented on esphome/components/mqtt/mqtt_backend_idf.cpp:
```diff
@@ -0,0 +1,149 @@
+
+  if (!this->lwt_topic_.empty()) {
+    mqtt_cfg_.lwt_topic = this->lwt_topic_.c_str();
+    // this->mqtt_cfg_.lwt_qos = this->lwt_qos_;
```
Should this be commented?

martgras replied:
fixed

oxan reviewed (commented):
Looks to be in pretty good shape now, I've just two questions remaining. 

I'm still not happy with the renaming of `mqtt_client.h`, but I'll leave that to @jesserockz to decide.

flaviut commented on esphome/components/mqtt/mqtt_backend_idf.cpp:
```diff
@@ -0,0 +1,149 @@
+  mqtt_cfg_.keepalive = this->keep_alive_;
+  mqtt_cfg_.disable_clean_session = !this->clean_session_;
+
+  if (this->username_.empty()) {
```
looks like the not disappeared here again.

martgras replied:
:) thanks

martgras committed:
address code review findings

martgras committed:
marks backend classes final

martgras committed:
lint again

martgras commented:
@jesserockz feel free to pick a different name for mqttclient.h/cpp and commit directly . I don't have a strong opinion on the name except that it shouldn't be mqtt_client to avoid the errors / confusion with the mqtt_client.h file from idf

oxan committed:
Revert mqttclient.h rename

oxan committed:
Merge branch 'dev' into mqtt-idf

martgras committed:
fix clang tidy

martgras committed:
Merge branch 'mqtt-idf' of https://github.com/martgras/esphome into mqtt-idf

jesserockz commented on esphome/components/mqtt/mqtt_client.h:
```diff
@@ -9,7 +9,11 @@
 #include "esphome/components/json/json_util.h"
 #include "esphome/components/network/ip_address.h"
-#include <AsyncMqttClient.h>
+#ifdef USE_ESP_IDF
+#include "mqtt_backend_idf.h"
+#else
+#include "mqtt_backend_arduino.h"
+#endif
```
Just for consistency
```suggestion
#if defined(USE_ESP_IDF)
#include "mqtt_backend_idf.h"
#elif defined(USE_ARDUINO)
#include "mqtt_backend_arduino.h"
#endif
```

jesserockz commented on esphome/components/mqtt/mqtt_client.h:
```diff
@@ -269,14 +268,20 @@ class MQTTClientComponent : public Component {
 
   std::vector<MQTTSubscription> subscriptions_;
-  AsyncMqttClient mqtt_client_;
+#ifdef USE_ESP_IDF
+  MQTTBackendIDF mqtt_backend_;
+#else
+  MQTTBackendArduino mqtt_backend_;
+#endif
```
Same as above
```suggestion
#if defined(USE_ESP_IDF)
  MQTTBackendIDF mqtt_backend_;
#elif defined(USE_ARDUINO)
  MQTTBackendArduino mqtt_backend_;
#endif
```

jesserockz reviewed (approved):
Mostly looks good to me.

martgras committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

Bascht74 commented:
Hi @martgras, do you have an idea if/when this PR will get merged or what is preventing that?
I am looking forward to use mqtt with esp-idf directly...

Seb

Bascht74 commented:
Hi @jesserockz, do you have an idea if/when this PR will get merged or what is preventing that?
I am looking forward to use mqtt with esp-idf directly...

Seb
